### PR TITLE
Substituted symbol equality

### PR DESCRIPTION
**Pull request by chsienki**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs [hunk 1]
```diff
@@ -498,6 +498,24 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
...
+            }
+
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            return ReferenceEquals(this.OriginalDefinition, other.OriginalDefinition) &&
+                   TypeSymbol.Equals(ContainingType, other.ContainingType, compareKind) &&
+                   TypeSymbol.Equals(Type, other.Type, compareKind);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs]: >TypeSymbol.Equals(Type, other.Type, compareKind) [](start = 19, length = 48)

This condition feels unnecessary.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 2]
```diff
@@ -1301,5 +1301,42 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
...
+            if (!TypeSymbol.Equals(this.ContainingType, other.ContainingType, compareKind)) return false;
+
+            // If both are declarations, then we don't need to check type arguments
+            // If exactly one is a declaration, then they re not equal
+            bool selfIsDeclaration = (object)this == (object)this.ConstructedFrom;
+            bool otherIsDeclaration = (object)other == (object)other.ConstructedFrom;
+            // PERF: VSadov specifically replaced the short-circuited operators in changeset #24717.
+            if (selfIsDeclaration | otherIsDeclaration)
+            {
+                return selfIsDeclaration & otherIsDeclaration;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs]: >return selfIsDeclaration & otherIsDeclaration; [](start = 16, length = 46)

This doesn't look right. The methods could be equal even when only one of them IsConstructedFrom

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 3]
```diff
@@ -1301,5 +1301,42 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
...
+            if (!ReferenceEquals(this.OriginalDefinition, other.OriginalDefinition))
+            {
+                return false;
+            }
+
+            // This checks if the methods have the same definition and the type parameters on the containing types have been
+            // substituted in the same way.
+            if (!TypeSymbol.Equals(this.ContainingType, other.ContainingType, compareKind)) return false;
+
+            // If both are declarations, then we don't need to check type arguments
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs]: >declarations [](start = 27, length = 12)

The term "declaration" is somewhat confusing here. Consider using "not constructed" instead. Same for variable names.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 4]
```diff
@@ -1301,5 +1301,42 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
...
+            if (selfIsDeclaration | otherIsDeclaration)
+            {
+                return selfIsDeclaration & otherIsDeclaration;
+            }
+
+            // This checks if the type parameters on the method itself have been substituted in the same way.
+            int arity = this.Arity;
+            for (int i = 0; i < arity; i++)
+            {
+                if (!this.TypeArgumentsWithAnnotations[i].Equals(other.TypeArgumentsWithAnnotations[i], compareKind))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs]: >this.TypeArgumentsWithAnnotations [](start = 21, length = 33)

Consider caching this value in a local.

File: src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs [hunk 5]
```diff
@@ -498,6 +498,24 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
             return visitor.VisitField(this);
         }
 
+        public override bool Equals(Symbol obj, TypeCompareKind compareKind)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs]: >override [](start = 15, length = 8)

sealed?

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs [hunk 6]
```diff
@@ -765,9 +959,12 @@ private void VerifyEquality(ISymbol type1, ISymbol type2, bool expectedDefault,
             Assert.Equal(expectedIncludeNullability, type2.Equals(type1, SymbolEqualityComparer.IncludeNullability));
 
             // GetHashCode
-            Assert.Equal(type1.GetHashCode(), type2.GetHashCode());
-            Assert.Equal(SymbolEqualityComparer.Default.GetHashCode(type1), SymbolEqualityComparer.Default.GetHashCode(type2));
-            Assert.Equal(SymbolEqualityComparer.IncludeNullability.GetHashCode(type1), SymbolEqualityComparer.IncludeNullability.GetHashCode(type2));
+            if (expectHashCodeMatch)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: >expectHashCodeMatch [](start = 16, length = 19)

Why is this conditional on a separate flag? If ```Object.Equals(Object)``` returns true, the hash code must match, so this should be conditioned on expectedDefault.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IFieldReferenceExpression.cs [hunk 7]
```diff
@@ -665,5 +667,50 @@ void M(C c1, C c2, int p1, int p2)
...
+                var model = comp.GetSemanticModel(syntaxTree);
+                var root = syntaxTree.GetRoot();
+                var classDecl = root.DescendantNodes().OfType<ClassDeclarationSyntax>().Single();
+                var classSym = (INamedTypeSymbol)model.GetDeclaredSymbol(classDecl);
+                var fieldSym = classSym.GetMembers("_field").Single();
+
+                var methodDecl = root.DescendantNodes().OfType<MethodDeclarationSyntax>().Single();
+                var methodBlockOperation = model.GetOperation(methodDecl);
+                var fieldReferenceOperation = methodBlockOperation.Descendants().OfType<IFieldReferenceOperation>().Single();
+                Assert.True(fieldSym.Equals(fieldReferenceOperation.Field));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IFieldReferenceExpression.cs]: >Assert.True(fieldSym.Equals(fieldReferenceOperation.Field)); [](start = 16, length = 60)

Please Assert that hash code is equal as well

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs [hunk 8]
```diff
@@ -744,7 +743,202 @@ public class B
...
+            Assert.IsType(typeof(SourceNamedTypeSymbol), event1ContainingType);
+            Assert.IsType(typeof(ConstructedNamedTypeSymbol), event2ContainingType);
+
+            VerifyEquality(event1ContainingType, event2ContainingType,
+                expectedDefault: true,
+                expectedIncludeNullability: false
+                );
+        }
+
+        private void VerifyEquality(ISymbol type1, ISymbol type2, bool expectedDefault, bool expectedIncludeNullability, bool expectHashCodeMatch = true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: >type1 [](start = 44, length = 5)

Please rename to symbol1 and the other to symbol2.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs [hunk 9]
```diff
@@ -744,7 +743,202 @@ public class B
...
+            var event2ContainingType = event2.ContainingType; //A<T!>
+
+            Assert.IsType(typeof(SourceNamedTypeSymbol), event1ContainingType);
+            Assert.IsType(typeof(ConstructedNamedTypeSymbol), event2ContainingType);
+
+            VerifyEquality(event1ContainingType, event2ContainingType,
+                expectedDefault: true,
+                expectedIncludeNullability: false
+                );
+        }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: >} [](start = 8, length = 1)

It looks like we are lacking tests for generic methods, matrix [(not constructed, constructed) method] * [(not constructed, constructed) containing type].

File: src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FieldSymbol.cs]: >Equals [](start = 29, length = 6)

What about events and properties?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

**Comment by chsienki** [comment]
Thanks @AlekseyTs. this isn't ready for review yet, but i'll keep the comments in mind as I update it

**Pull request closed without merging**